src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.Entry.Roundtrip.Tests.cs

```diff
@@ -0,0 +1,137 @@
+            Assert.Equal(name, entry.Name);
+        }
+
+        public static IEnumerable<object[]> LinkNameRoundtripsTheoryData()
+        {
+            foreach (TarEntryType entryType in new[] { TarEntryType.SymbolicLink, TarEntryType.HardLink })
+            {
+                foreach (string name in GetNamesNonAsciiTestData(NameCapabilities.Name).Concat(GetNamesPrefixedTestData(NameCapabilities.Name)))
+                {
+                    yield return new object[] { TarEntryFormat.V7, entryType, name };
+                    yield return new object[] { TarEntryFormat.Ustar, entryType, name };
+                }
+
+                foreach (string name in GetNamesNonAsciiTestData(NameCapabilities.Unlimited).Concat(GetNamesPrefixedTestData(NameCapabilities.Unlimited)))
+                {
+                    yield return new object[] { TarEntryFormat.Pax, entryType, name };
+                    yield return new object[] { TarEntryFormat.Gnu, entryType, name };
+                }
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(LinkNameRoundtripsTheoryData))]
+        public void LinkNameRoundtrips(TarEntryFormat entryFormat, TarEntryType entryType, string linkName)
+        {
+            string name = "foo";
+            TarEntry entry = InvokeTarEntryCreationConstructor(entryFormat, entryType, name);
+            entry.LinkName = linkName;
+
+            MemoryStream ms = new();
+            using (TarWriter writer = new(ms, leaveOpen: true))
+            {
+                writer.WriteEntry(entry);
+            }
+
+            ms.Position = 0;
+            using TarReader reader = new(ms);
+
+            entry = reader.GetNextEntry();
+            Assert.Null(reader.GetNextEntry());
+            Assert.Equal(name, entry.Name);
+            Assert.Equal(linkName, entry.LinkName);
+        }
+
+        public static IEnumerable<object[]> UserNameGroupNameRoundtripsTheoryData()
+        {
+            foreach (TarEntryFormat entryFormat in new[] { TarEntryFormat.Ustar, TarEntryFormat.Pax, TarEntryFormat.Gnu })
+            {
+                yield return new object[] { entryFormat, Repeat(OneByteCharacter, 32) };
+                yield return new object[] { entryFormat, Repeat(TwoBytesCharacter, 32 / 2) };
+                yield return new object[] { entryFormat, Repeat(FourBytesCharacter, 32 / 4) };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(UserNameGroupNameRoundtripsTheoryData))]
+        public void UserNameGroupNameRoundtrips(TarEntryFormat entryFormat, string userGroupName)
+        {
+            string name = "foo";
+            TarEntry entry = InvokeTarEntryCreationConstructor(entryFormat, TarEntryType.RegularFile, name);
+            PosixTarEntry posixEntry = Assert.IsAssignableFrom<PosixTarEntry>(entry);
+            posixEntry.UserName = userGroupName;
+            posixEntry.GroupName = userGroupName;
+
+            MemoryStream ms = new();
```

Comment:
Note to myself: The [fix for extended attributes](https://github.com/dotnet/runtime/pull/75237) should include this test and extend it to an unseekable stream archive, since the uname and gname get written into the data section.